(#478) Lazy post content loading + youtube titles and durations parsing + unseen posts indicator + inlined files' size fetching

K1rakishou opened the pull request:
Closes #478, #526, #460, #779, #834, #837

- [x] Media prefetching works via lazy loading.
- [x] #478 - Youtube titles/durations fetching works via lazy loading. (https://files.catbox.moe/f2sp8g.mp4)
- [x] DB cache for titles/durations. (New Database was introduced - Room. I'm planning to gradually migrate our old DB (ormlite) to Room because Room is times better).
- [x] #526 Seen/unseen posts indicator. (https://files.catbox.moe/4rjv3w.mp4)
- [x] #460 Maybe load inlined files' info via HEAD requests too? (this one is kinda bad for security (because of ip loggers and shit) BUT you can already open a file with an ip logger just by opening an inlined file). On the other hand, it would be a nice addition because inlined files may take hundreds of megabytes and you won't even know until opening it. Should be disabled by default and have a huge warning on it.
- [x] Code refactoring.
- [x] Settings to enable and disable all of those things.
- [x] Don't forget to fix this thing: https://github.com/Adamantcheese/Kuroba/issues/779
- [x] Don't forget to add a setting to only show title/duration without the link itself.
- [x] ~~Add ability to disable seen posts label fading out animation. And improve it a little bit (right now if an unseen post has a youtube link or something like that, that will call notifyItemChanged on that post, the label will be hidden right after the update). In the end I decided to remove animation completely.~~ Actually no, I'm planning to bring back the animation. And it's already back.
- [x] More unit tests.
- [x] Testing.
    - [x] Check that spoilered youtube links work as expected.
    - [x] Make sure that enabling and disabling of each feature works as expected.
    - [x] Make sure that everything works in local threads.

Bugs to fix:
- [x] Also, there is a bug: the last scroll position in reply window is not restored when you click on any inner reply and then back: you get scrolled to top. Should remember the scroll position and restore it.
- [x] For some reason adds youtube title spans twice to the same link.
- [x] Youtube titles/durations fetching with 4chanx's API key, when it exceeded it's quota, doesn't work as expected.
- [x] Unseen posts' label is kinda broken. Right now it will disappear right after all loaders have done their job. 

Debug apk for anyone interested: 
[Kuroba-abe472e.zip](https://github.com/Adamantcheese/Kuroba/files/4364939/Kuroba-abe472e.zip)


The app will probably crash when going back to regular dev version from this one (because the database version was changed) so **don't forget to export settings before installing this apk**.

K1rakishou committed:
(#478) First prototype of OnDemandContentLoaderManager

- Implement OnDemandContentLoaderManager + content loader prototype.
- Remove all onViewAttachedToWindow/onViewDetachedFromWindow in PostCell (and other Post classes) and PostAdapter. The reasoning is it's bad to use those two callbacks with RecyclerView because they
have separate lifecycles. So when using onViewAttachedToWindow/onViewDetachedFromWindow you may end up in a situation when onViewDetachedFromWindow was called, you nulled out all the callbacks but
for RecyclerView that we might still be alive so you will end up with either the NPE or empty view.
- Replace ListView with RecyclerView in PostRepliesController because ListView would sometimes bind the same view twice which would break content loading subscription/unsubscription. And there was no
way to manually recycle replies when closing that controller.

K1rakishou committed:
(#478) CacheHandler fixes + a little bit of refactoring (extracted some classes into their own files)

K1rakishou committed:
(#478) SmartPrefetch implementation

Prefetch now won't start prefetching every single media file in a thread/catalog as soon as you open it. Instead, the post media will now only start prefetching in 1 second after post binding.
If during that time frame the post was unbound the prefetch is canceled.

K1rakishou committed:
(#478) Youtube link title and durations parsing via on-demand content loading draft

K1rakishou committed:
(#478) Can now load youtube titles and durations via on-demand content loading!

K1rakishou committed:
(#478) Refactor youtube link info fetcher into it's own separate class + make it easy to add new fetchers (let's say for SoundCloud/Vimeo/DailyMotion etc.)

- Make sure that when there are multiple links that are the same in one post only ONE network request is sent.
- Add icon for youtube links.
- Log all unhandled errors in loaders and in the on-demand loader manager too.
- Update kotlin to 1.3.70

K1rakishou committed:
(#478) Merge branch 'cheese-multi-feature' into (#478)-on-demand-post-content-loading

# Conflicts:
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/presenter/ThreadPresenter.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/controller/PostRepliesController.java

K1rakishou committed:
(#478) Add DB cache for post extra content

- New DB, Room, that lives in a separate module. In future our old DB (ormlite) will be gradually moved to Room and then removed from the project.

K1rakishou committed:
(#478) Database refactoring + fix bug when one post contains multiple youtube links and only the first one will be stored in the DB + store original video url instead of url to get video info (this should be changed to videoId in the future) + introduce LoadableEntity

K1rakishou committed:
(#478) Introduce a label for not yet seen posts

K1rakishou committed:
(#478) Merge branch 'cheese-multi-feature' into (#478)-on-demand-post-content-loading

# Conflicts:
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/di/ManagerModule.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/cell/PostCell.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/utils/KtExtensions.kt

K1rakishou committed:
(#478) Post merge fixes

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/loader/impl/PrefetchLoader.kt:
```diff
@@ -0,0 +1,114 @@
+            PostImage.Type.GIF -> shouldLoadForNetworkType(ChanSettings.imageAutoLoadNetwork.get())
+            PostImage.Type.MOVIE -> shouldLoadForNetworkType(ChanSettings.videoAutoLoadNetwork.get())
+            PostImage.Type.PDF,
+            PostImage.Type.SWF -> false
```
I think I need to also handle mp3/swf and other new supported types. By handling I mean filtering them out.

K1rakishou replied:
Actually they all are a part of the MOVIE type which is already handled.

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/loader/impl/external_media_service/ExternalMediaServiceExtraInfoFetcher.kt:
```diff
@@ -0,0 +1,50 @@
+    ): Flowable<ModularResult<Unit>>
+
+    /**
+     * Whether a link belongs to this fetcher
```
"Whether this fetcher can parse the link."

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/loader/impl/external_media_service/YoutubeMediaServiceExtraInfoFetcher.kt:
```diff
@@ -0,0 +1,191 @@
+                PostExtraContentLoader.formatterWithoutHours.print(time)
+            }
+        } catch (ignored: Throwable) {
+            null
```
Maybe I should log something here.

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/loader/impl/external_media_service/YoutubeMediaServiceExtraInfoFetcher.kt:
```diff
@@ -0,0 +1,191 @@
+                    .get("title")
+                    .asString
+        } catch (ignored: Throwable) {
+            null
```
Maybe I should log something here.

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/loader/impl/post_comment/CommentSpanUpdater.kt:
```diff
@@ -0,0 +1,161 @@
+                )
+
+                if (originalLinkUrl == formattedLinkUrl) {
+                    return@forEach
```
Probably should only add the icon here.

K1rakishou replied:
Actually this check is pointless since I always add two spaces "  " to every link to prepend the icon.

K1rakishou commented on Kuroba/database/src/main/java/com/github/adamantcheese/database/KurobaDatabase.kt:
```diff
@@ -0,0 +1,44 @@
+            YoutubeLinkExtraContentEntity::class,
+            SeenPostEntity::class
+        ],
+        version = 1
```
Add schema export.

K1rakishou commented:
~~Also, there is a bug: the last scroll position in reply window is not restored when you click on any inner reply and then back: you get scrolled to top. Should remember the scroll position and restore it.~~

K1rakishou commented on Kuroba/database/src/main/java/com/github/adamantcheese/database/di/annotation/VerboseLogs.kt:
```diff
@@ -0,0 +1,4 @@
+package com.github.adamantcheese.database.di.annotation
+
+@Retention(value = AnnotationRetention.SOURCE)
```
Add Qualifier annotation.

K1rakishou commented on Kuroba/database/src/main/java/com/github/adamantcheese/database/di/annotation/LoggerTagPrefix.kt:
```diff
@@ -0,0 +1,4 @@
+package com.github.adamantcheese.database.di.annotation
+
+@Retention(value = AnnotationRetention.SOURCE)
```
Add Qualifier annotation.

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/Chan.java:
```diff
@@ -105,13 +110,23 @@ public void onCreate() {
+
         feather = Feather.with(
                 new AppModule(this),
+                new ExecutorsManager(),
```
ExecutorsModule

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/di/LoaderModule.java:
```diff
@@ -0,0 +1,54 @@
+    @Provides
+    @Singleton
+    public PrefetchLoader providePrefetchLoader(FileCacheV2 fileCacheV2) {
+        return new PrefetchLoader(fileCacheV2);
```
Logging (AppModule.DI_TAG)

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/di/LoaderModule.java:
```diff
@@ -0,0 +1,54 @@
+    public YoutubeMediaServiceExtraInfoFetcher provideYoutubeMediaServiceExtraInfoFetcher(
+            YoutubeLinkExtraContentRepository youtubeLinkExtraContentRepository
+    ) {
+        return new YoutubeMediaServiceExtraInfoFetcher(youtubeLinkExtraContentRepository);
```
Logging (AppModule.DI_TAG)

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/di/LoaderModule.java:
```diff
@@ -0,0 +1,54 @@
+            YoutubeMediaServiceExtraInfoFetcher youtubeMediaServiceExtraInfoFetcher,
+            @Named(ExecutorsManager.onDemandContentLoaderExecutorName) Executor onDemandContentLoaderExecutor
+    ) {
+        List<ExternalMediaServiceExtraInfoFetcher> fetchers = new ArrayList<>();
```
Logging (AppModule.DI_TAG)

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/di/ExecutorsManager.java:
```diff
@@ -0,0 +1,45 @@
+    @Singleton
+    @Named(onDemandContentLoaderExecutorName)
+    public Executor provideOnDemandContentLoaderExecutor() {
+        return createExecutor(
```
Logging (AppModule.DI_TAG)

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/loader/LoaderResult.kt:
```diff
@@ -0,0 +1,21 @@
+     * network and the loader requires Wi-Fi connection to load huge content, like prefetcher).
+     * When Rejected is returned that means that we don't need to update the post (there is no info)
+     * */
+    class Rejected(loaderType: LoaderType) : LoaderResult(loaderType)
```
I think these three should be renamed into Succeeded, Failed and Rejected.

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/loader/LoaderType.kt:
```diff
@@ -0,0 +1,11 @@
+    PostExtraContentLoader,
+
+    // TODO: delete
+    YoutubeLinkTitlesLoader,
```
Don't forget to delete unused enums.

K1rakishou commented:
~~Check that spoilered youtube links work as expected.~~

K1rakishou committed:
(#478) Code review fixes

K1rakishou committed:
(#478) Merge branch 'cheese-multi-feature' into (#478)-on-demand-post-content-loading

# Conflicts:
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/Chan.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/presenter/ThreadPresenter.java

K1rakishou committed:
(#478) Add fade out animation for unseed post attention label

K1rakishou committed:
(#478) Merge branch 'cheese-multi-feature' into (#478)-on-demand-post-content-loading

# Conflicts:
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/cache/FileCacheV2.kt
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/settings/SettingProvider.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/settings/SharedPreferencesSettingProvider.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/settings/json/JsonSettingsProvider.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/parser/CommentParserHelper.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/cell/CardPostCell.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/cell/PostCell.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/cell/PostCellInterface.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/controller/settings/ThemeSettingsController.java

K1rakishou committed:
(#478) Post merge fixes

K1rakishou committed:
(#478) Increase animation delay and duration

K1rakishou committed:
(#478) Refactoring

YoutubeLinkExtraContent -> MediaServiceLinkExtraContent

K1rakishou committed:
(#478) More refactoring

Move all of the get/store code from YoutubeMediaServiceExtraInfoFetcher into MediaServiceLinkExtraContentRepository
Add MediaServiceLinkExtraContentRemoteSource
Rename database module into model

K1rakishou committed:
(#478) Implement InlinedFileInfoLoader (plus DB caching)

K1rakishou committed:
(#478) Fix prefetches getting canceled once we call notifyItemChanged in PostAdapter + refactoring

K1rakishou committed:
(#478) Fix a bug where if a prefetch was started and then canceled before it managed to download an image then on the next time a post is bound we won't attempt to prefetch images for that post again because
of the global (for a single Post) flag onDemandContentLoaded

Rewrite SeenPostsManager to use actors

K1rakishou committed:
(#478) Store and restore reply window scroll position (which was broken after switching to RecyclerView)

K1rakishou committed:
(#478) Re-apply back old spans that were on top a PostLinkable span (with updated spanEnds)

K1rakishou committed:
(#478) Fix comment

K1rakishou committed:
(#478) Merge branch 'cheese-multi-feature' into (#478)-on-demand-post-content-loading

# Conflicts:
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/parser/CommentParserHelper.java

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/loader/impl/external_media_service/ExternalMediaServiceExtraInfoFetcher.kt:
```diff
@@ -0,0 +1,33 @@
+
+    fun isEnabled(): Boolean
+
+    fun fetch(requestUrl: String, linkInfoRequest: LinkInfoRequest): Flowable<ModularResult<SpanUpdateBatch>>
```
I think this method should return Single because it's expected to return only one result and Flowable may be used to return a stream of results which would probably break things.

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/cell/PostCell.java:
```diff
@@ -352,6 +374,73 @@ private void bindPost(Theme theme, Post post) {
+    }
+
+    private void startAttentionLabelFadeOutAnimation() {
+        if (hasColoredFilter) {
```
Check that postAttentionLabel is visible.

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/layout/ThreadListLayout.java:
```diff
@@ -151,13 +152,13 @@ public void setCallbacks(
-            recyclerView.getRecycledViewPool().setMaxRecycledViews(TYPE_POST, 0);
-        }
+        // Man, fuck the RecycledViewPool. Sometimes when scrolling away from a view and the swiftly
+        // back to it onViewRecycled() will be called TWICE for that view. Setting setMaxRecycledViews
```
Setting setMaxRecycledViews for TYPE_POST to 0

K1rakishou commented on Kuroba/model/build.gradle:
```diff
@@ -0,0 +1,55 @@
+    implementation 'androidx.core:core-ktx:1.2.0'
+    implementation 'joda-time:joda-time:2.10.5'
+    implementation "org.jetbrains.kotlinx:kotlinx-coroutines-core:${versions.kotlin_coroutines_version}"
+    implementation 'com.squareup.okhttp3:okhttp:4.4.0'
```
Extract all of these into versions

K1rakishou commented on Kuroba/model/src/main/java/com/github/adamantcheese/database/data/loadable/LoadableUtils.kt:
```diff
@@ -0,0 +1,24 @@
+
+import java.util.*
+
+object LoadableUtils {
```
Should try to extract this into the base module.

K1rakishou commented on Kuroba/model/src/main/java/com/github/adamantcheese/database/repository/SeenPostRepository.kt:
```diff
@@ -0,0 +1,50 @@
+
+    suspend fun insert(seenPost: SeenPost): ModularResult<Unit> {
+        return runInTransaction {
+            seenPostLocalRepositoryCleanup().unwrap()
```
Should probably, too, use ignore() here.

K1rakishou commented on Kuroba/model/src/main/java/com/github/adamantcheese/database/source/remote/AbstractRemoteSource.kt:
```diff
@@ -0,0 +1,34 @@
+
+    protected suspend fun OkHttpClient.suspendCall(request: Request): Response {
+        return suspendCancellableCoroutine { continuation ->
+            newCall(request).enqueue(object : Callback {
```
Add request cancellation. Right now it's uncancellable.

K1rakishou committed:
(#478) Merge branch 'cheese-multi-feature' into (#478)-on-demand-post-content-loading

K1rakishou committed:
(#478) Code review fixes

K1rakishou committed:
(#478) Fix a bug where an unavailable youtube video (deleted or something else) would not even render youtube icon. Now if a video title and duration (both) are not available then we will append [NOT AVAILABLE] at the end of the link and if some kind of unknown error had occurred while trying to fetch video's title/duration then we will append [ERROR].

K1rakishou committed:
(#478) Add "dependsOnSetting" for BooleanSettingView which will help to build a "depends-on" relational model for boolean setting (i.e. automatically disable a child setting if it's parent is disabled) + fix an NPE in Post.Builder() when opening ThemeSettingsController

K1rakishou committed:
(#478) Add new settings + remove some old, unused settings

- Add settings to enable/disable inlined files' size fetching, unseen posts marking.
- Merge parseYoutubeTitles and parseYoutubeDuration into one setting (parseYoutubeTitlesAndDuration).
- Remove parse_youtube_titles, parse_youtube_duration and yt_cache from SharedPreferences.

K1rakishou committed:
(#478) Add an explanation for why a mutex is used inside of an actor

K1rakishou committed:
(#478) Fix a bug where a post with youtube link wouldn't get updated, after fetching it's title and duration, when lastSeenIndicatorPosition is visible. That happened because of miscalculation of the post's index due to lastSeenIndicatorPosition

K1rakishou committed:
(#478) Add "isInlined" parameter to SelializablePostImage so that inlined files work in local threads

K1rakishou committed:
(#478) Fix couple of issues related to loading inlined media in local threads (where we would attempt to load it from the disk instead of the network)

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/di/LoaderModule.java:
```diff
@@ -0,0 +1,73 @@
+            InlinedFileInfoRepository inlinedFileInfoRepository,
+            @Named(ExecutorsModule.onDemandContentLoaderExecutorName) Executor onDemandContentLoaderExecutor
+    ) {
+        return new InlinedFileInfoLoader(
```
Add logging AppModule.DI_TAG

K1rakishou committed:
(#478) Add couple of comments

K1rakishou committed:
(#478) Merge branch 'cheese-multi-feature' into (#478)-on-demand-post-content-loading

# Conflicts:
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/cell/PostCell.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/layout/ThreadListLayout.java

K1rakishou committed:
(#478) Merge with dev

K1rakishou committed:
(#478) Fix some PostLinkable spans getting doubled on some sites (like 2ch.hk)

K1rakishou committed:
(#478) Fix a memory leak in PostNumberClickableSpan and another one in the new PostCell animation

K1rakishou committed:
(#478) Merge branch 'cheese-multi-feature' into (#478)-on-demand-post-content-loading

# Conflicts:
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/controller/settings/ThemeSettingsController.java

K1rakishou committed:
(#478) Merge with dev

K1rakishou committed:
(#478) Remove unseen posts indicator fade out animation

K1rakishou committed:
(#478) Add a setting to disable showing the original youtube link and with the video title and duration (like how it works in 4chanx). Fix a bug in dependOnSetting where it would always have the false state by default.

K1rakishou committed:
(#478) Implement caches for InlinedFileInfoRepository and MediaServiceLinkExtraContentRepository

K1rakishou committed:
(#478) We now store videoId instead of the whole link for Youtube videos + we now store raw period of time instead of already formatted one.

K1rakishou committed:
(#478) Implement tests for InlinedFileInfoLocalSource/InlinedFileInfoRemoteSource/InlinedFileInfoRepository

K1rakishou committed:
(#478) Implement tests for MediaServiceLinkExtraContentLocalSource/MediaServiceLinkExtraContentRemoteSourceTest

Refactor some repositories to share code for the most common operation

K1rakishou commented:
Alright, I think this PR is ready for review. Only a couple of classes are left to cover them with unit-tests and the general testing is pretty much done. But do not merge yet.

K1rakishou committed:
(#478) Module names refactoring:

base -> core-common
model -> core-model

K1rakishou committed:
(#478) Add ensureBackgroundThread() and ensureMainThread() checks where necessary

K1rakishou committed:
(#478) Extract filters into PostFilter object

K1rakishou committed:
(#478) Pass OkHttpDns and OkHttpProtocols into the core-model module

K1rakishou committed:
(#478) Implement tests for MediaServiceLinkExtraContentRepository, fix other tests

K1rakishou committed:
(#478) Add a fix for CacheHandler when a cache meta size becomes negative

K1rakishou committed:
(#478) Finally fix the leak in FastTextView.textCache

K1rakishou committed:
(#478) Revert "(#478) Remove unseen posts indicator fade out animation"

This reverts commit b41c0e29

K1rakishou committed:
(#478) Only hide the unseen post label after all loaders have completed loading

K1rakishou committed:
(#478) Hide the unseen post label completely after the alpha animation has completed

K1rakishou committed:
(#478) Rename bunch of stuff + fixes in comments

K1rakishou committed:
(#478) Fix some warnings in tests

K1rakishou committed:
(#478) Remove old database schema file

K1rakishou committed:
(#478) Update travis.yml so that the CI always runs tests

K1rakishou committed:
(#478) Get rid of more warnings

K1rakishou committed:
(#478) Merge branch 'cheese-multi-feature' into (#478)-on-demand-post-content-loading

# Conflicts:
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/manager/FilterEngine.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/mapper/PostImageMapper.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/model/PostImage.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/presenter/ThreadPresenter.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/parser/CommentParserHelper.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/cell/PostCell.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/controller/settings/ThemeSettingsController.java

K1rakishou committed:
(#478) Merge with dev

K1rakishou committed:
(#478) Only add delay for loaders if one of the loaders has no cached data for a post + add a loading indicator for prefetching images

K1rakishou committed:
(#478) Increase the loader loading delay a little bit

K1rakishou committed:
(#478) Fix unseen posts indicator would be hidden right away a post is updated via notifyItemChanged (this happens when we update a post after all loaders have done their work)

K1rakishou committed:
(#478) Fix OnDemandContentLoaderManager and add one more test

K1rakishou committed:
(#478) A little bit of improvements for OnDemandContentLoaderManager tests

K1rakishou committed:
(#478) Fix a crash when trying to clear SingletonList + add more logs to LocalSources

K1rakishou committed:
(#478) Comments/strings and other tiny fixes

K1rakishou committed:
(#478) Merge branch 'cheese-multi-feature' into (#478)-on-demand-post-content-loading

K1rakishou committed:
(#478) Fix thumbnails' previews appearance animation not working properly sometimes thus causing them to stay slightly grayish

Separate PostImageThumbnailView.setPostImage into bindPostImage() and unbindPostImage()

K1rakishou committed:
(#478) Fix bitmap leak in ThumbnailView + hide prefetch loading indicator for album thumbnails

K1rakishou committed:
(#478) Do not log when we delete files which size is 0.
Fix a bug where if youtube titles/durations fetching is turned off the youtube link would only contain the icon (without the link itself). Basically it would be empty.
Fix (hopefully) memory leak in PostImageThumbnailView.

K1rakishou committed:
(#478) Remove everything related to Loadable2

K1rakishou commented:
So I think I'm gonna close this PR to avoid having to write a huge ass migration for the Room database since I'm moving the Ormlite database into the Room.

K1rakishou committed:
(#478) Merge branch 'cheese-multi-feature' into (#478)-on-demand-post-content-loading

# Conflicts:
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/model/Post.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/parser/CommentParserHelper.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/cell/PostCell.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/layout/ThreadLayout.java
#	Kuroba/app/src/main/res/values/strings.xml
#	Kuroba/build.gradle

K1rakishou committed:
(#478) Merge with dev

K1rakishou committed:
(#478) Remove logging

K1rakishou commented:
In case you are wondering what happened to me refactoring the whole app - I gave up because I attempted to literally rewrite like half of the app during the refactoring and it ended up in me realizing that it will take me months to only do that (let alone testing the WHOLE app from scratch because the amount of bugs was ridiculous). 

So fuck it. It was fun but it's impracticable. It's easier to just rewrite everything from scratch.

Adamantcheese commented:
If you want, you can run your own fork of the application while you slowly work on porting it over to Kotlin. It is set up for that after all.

K1rakishou commented:
Nah, it's really too much work. I spent past 3 weeks on it and I managed to convert 70% of the code into Kotlin, fix most of the places where we access the database on the main thread (changed them to use coroutines) and even moved some of the tables into Room (Loadable, Board, Filter, SavedBoard, PostHide). I almost made it so that the database is the primary source of truth, meaning if you want to change something (say loadable) the only way to do that is to also change it in the database as well. When you read a loadable from the database it will return you an immutable object, so you have to use the `update` function that will actually give you a mutable object (it provides a lambda where you can do all the mutations). Also I managed to change all of the places where we use Loadables as keys for Maps and stuff (because using mutable keys is very bad) and even more - I changed most of the places where we used loadable to a new type of object - LoadableDescriptor. Basically it's an object that has three fields - site name, board code and thread id (If we want to get a catalog then thread id will be null). It's immutable and it's being used as a key to get the real loadable from the database when you need it (Most of the time you don't need it, though). And I did the same thing for boards and sites too (BoardDescriptor and SiteDescriptor).

But then I decided, for some reason, to also get rid of java.util.Observable and ended up in a situation where I will have to rewrite site/board setup screens with all the logic almost from scratch (I did unwrapped some circular dependencies and removed some of the classes that were really unnecessary). And then there is also testing. I will literally have to re-test every single action in the app because of how broken everything has become. I even started writing UI tests but I will have to write hundreds of them. 

So it's too much work. It was fine while I was on vacation, but now I can only work on it on weekends so it will take even more time.

K1rakishou committed:
(#478) Change the database a little bit

Get rid of loadableUid/postUid, use site name/board code as a catalog identifier instead

K1rakishou committed:
(#478) Database optimizations + some fixes for 2ch.hk

Change the DB, add ChanThread column (which depends on ChanBoard column).
Possible fix for "Thread has no op" error since "op" parameter is not reliable for determining whether a post is OP or not (we now use "parent" parameter which is always 0 for OP and opId for every other post in a thread).

K1rakishou committed:
(#478) Change MockReplyManager to use new ThreadDescriptor, remove the old one

K1rakishou committed:
(#478) Fix tests

K1rakishou committed:
(#478) Port DB fixes from another branch to this one

K1rakishou commented:
Fuck, something broke and now the database in my another branch, where I'm working on archives, doesn't like the old version of itself from this branch and requires me to write a migration even though they should be the same. I have no idea how to fix this without breaking the database along the way so I'm gonna close this PR so I don't have to write a migration whatsoever. (My other branch with archives is based on this one anyway).
